include/deal.II/trilinos/nox.h

```diff
@@ -0,0 +1,1029 @@
+              }
+            else
+              {
+                const auto &x = problem.getSolutionGroup().getX();
+                const auto *x_ =
+                  dynamic_cast<const internal::Vector<VectorType> *>(&x);
+
+                const auto &f = problem.getSolutionGroup().getF();
+                const auto *f_ =
+                  dynamic_cast<const internal::Vector<VectorType> *>(&f);
+
+                const unsigned int step = problem.getNumIterations();
+
+                const double norm_f = f_->genericVector().l2_norm();
+
+                state = this->check_iteration_status(step,
+                                                     norm_f,
+                                                     x_->genericVector(),
+                                                     f_->genericVector());
+
+                switch (state)
+                  {
+                    case SolverControl::iterate:
+                      status = NOX::StatusTest::Unconverged;
+                      break;
+                    case SolverControl::failure:
+                      status = NOX::StatusTest::Failed;
+                      break;
+                    case SolverControl::success:
+                      status = NOX::StatusTest::Converged;
+                      break;
+                    default:
+                      AssertThrow(false, ExcNotImplemented());
+                  }
+              }
+          }
+
+        if (as_dummy)
+          status = NOX::StatusTest::Unconverged;
+
+        return status;
+      }
+
+      NOX::StatusTest::StatusType
+      getStatus() const override
+      {
+        return status;
+      }
+
+      virtual std::ostream &
+      print(std::ostream &stream, int indent = 0) const override
+      {
+        (void)indent;
+
+        std::string state_str;
+        switch (state)
+          {
+            case SolverControl::iterate:
+              state_str = "iterate";
+              break;
+            case SolverControl::failure:
+              state_str = "failure";
+              break;
+            case SolverControl::success:
+              state_str = "success";
+              break;
+            default:
+              AssertThrow(false, ExcNotImplemented());
+          }
+
+        for (int j = 0; j < indent; j++)
+          stream << ' ';
+        stream << status;
+        stream << "check_iteration_status() = " << state_str
+               << " (dummy = " << (as_dummy ? "yes" : "no") << ")";
+        stream << std::endl;
+
+        return stream;
+      }
+
+    private:
+      std::function<SolverControl::State(const unsigned int,
+                                         const double,
+                                         const VectorType &,
+                                         const VectorType &)>
+        check_iteration_status = {};
+
+      const bool as_dummy = false;
+
+      NOX::StatusTest::StatusType status;
+      SolverControl::State        state;
```

Comment:
Can you document these member variables?

I will note that it is quite confusing to have both `status` and `state` variables and would much appreciate if they were more clearly named.